README.md
<div>
  <br/>
  <br/>
  <p>
  	<a href="https://travis-ci.com/theqoobee/rpcengine"><img src="https://travis-ci.com/theqoobee/rpcengine.svg?branch=master"></a>
    <a href="https://www.npmjs.com/package/discord.js"><img src="https://img.shields.io/npm/v/discord.js.svg?maxAge=3600" alt="NPM version" /></a>
    <a href="https://david-dm.org/theqoobee/rpcengine"><img src="https://img.shields.io/david/theqoobee/rpcengine.svg?maxAge=3600" alt="Dependencies" /></a>
    <a href="https://github.com/theqoobee/rpcengine/releases"><img src="https://img.shields.io/github/v/release/theqoobee/rpcengine" /></a>
  </p>

</div>


# RPC | Engine
A simple application to set custom Discord Rich Presence statuses!



### __What is it?__

Discord Rich Presence, or RPC for short, is a way to show off what you're doing on your computer inside discord. 
Normal rich presence is **not** customizable, and that's the purpose of **rpcengine**.

**Rpcengine** fills the gap discord usual **rich presence** leaves open, and provides an easy way to create **custom** and **fully editable** rich presence statuses. Just install the app, run it and start flexing on people too lazy to google.

***

<img width="250px" src="https://github.com/theqoobee/rpcengine/blob/master/icons/therpcimage.png?raw=true#thumbnail">

<img width="700px" src="https://github.com/theqoobee/rpcengine/blob/master/icons/semishort_rounded.png?raw=true">

***

## Getting Started

Rpcengine is made in a way that basically anyone can use it and have a good time. There are only a couple different settings to configure. To view all the additional settings go to the [usage page](https://github.com/theqoobee/rpcengine/wiki/usage)

### Installation

Download the installer for your platform from [here](https://github.com/theqoobee/rpcengine/releases) and follow the very simple instructions to install it!  

## List of features 
- Completely customize your discord rich presence, as seen in the images above
- Add an optional timer to show how long you've been active
- Set images 

Optionally set any of the following images in your richpresence
**(you can add custom images, and these are only some of the available ones)**
```
Android Studio
Apple
Arch Linux
Atom
Cat
Chibi
Chrome
C++
Darkness
Deku
Discord
Electron
Elm
Facebook
Github
Gitlab
Google
Google Music
Instagram
Intellij IDEA
Java
Javascript
Kubernetes
Linux
Megu
Megu1
Music
Naruto sleep
Naruto small 
Node.js
Php
Poopy
Python
Rem
Ruby
Sasuke
Shell
Swift
Tanjiro
Todoroki
Visual Studio
VSCode
Webstorm
Windows
XCode
Youtube


## License
[MIT](https://choosealicense.com/licenses/mit/)

changelog.md
# Changelog
All notable changes to this project will be documented in this file.

(The github releases only periodically get full fledged finished versions)

## [Unreleased]

## [2.1.0] - 2020-01-27
### Added
- An initiative to actually keep a changelog
- Dynamic image options.
- Custom client Ids
- Custom images
- Support for large image texts and small image texts
- Added the changelog (omg)
- Added a wiki page
- Created a settings window
- Started UI refresh

### Changed
- Fixed minor bugs in the old version.
- Rewrote parts of `rpc.js`
- Improved documentation and commenting

### Removed
- Needless options in `index.html` and started dynamically adding them.

## [2.0.0] - 2020-1-23
### Added
- Better documentation
- Linting
- Build checking
- Started properly using `javascript`
- Using standard 

### Changed
- Literally rewrote **everything**
- Top down redesign of the entire system

## [1.5.0] - 2020-1-17
### Changed
- Pretty up the code a little bit
- Remove unnecessary stuff
 
## [Released]
## [1.0.0] - 2020-1-15
### Added
- Code.

### Changed
- Turned an empty repo into a not-so-empty one



src/js/rpc.js
// MIT License

// Copyright (c) 2020 theqoobee, qoobes

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

const DiscordRPC = require('discord-rpc') // importing the discord library

var clientId = '665244956805300274' // The clientId of the discord app
const ipcRPC = require('electron').ipcRenderer // This is how we send messages between this and main.js
const timestamp = new Date() // This is what we use for the timestamp
var username
// Default vars

var tempArgs
var args = {
  details: 'Using rpcengine',
  state: 'by theqoobee',
  startTimestamp: true, // Timestamp to track the passage of time
  largeImageKey: 'megu1', // name of the pics
  smallImageKey: 'naruto_sleep', // ^^^
  largeImageText: "I'm big image", // the text displayed when hovering over them
  smallImageText: "I'm small image",
  instance: false
}

var rpc = new DiscordRPC.Client({ transport: 'ipc' }) // setting up the richpresence object
// The purpose of the following function is to check whether or not there are any updates to the arguments
//
// It sends a synchornous message to main via ipcRenderer, and as a result gets either false, meaning there
// were no changes, or a changed version of the arguments
ipcRenderer.send('info')
async function check () {
  var update = ipcRPC.sendSync('check', username) // passing the username as an argument
  if (update.args) { // If back is true, in other words if it isn't false, null, or undefined
    args = update.args // change the value of args to the value given by the main process
  }
  if (update.clientId) {
    clientId = update.clientId
    clientUpdate()
  }
}

// The function to set the rich presence of the user
//
async function setActivity () {
  if (!rpc) { // if the rpc object was not properly initialized, do not run
    return
  }
  // here i chose to make a temporary version of the arguments for the sake of not changing
  // the real ones when i set the timestamp there because the user turning them off would be a problem
  tempArgs = args
  if (args.startTimestamp) { // if the user chose to have a timestamp
    tempArgs.startTimestamp = timestamp // set the field to a timestamp
  } else { tempArgs.startTimestamp = null } // else set it to null

  // send the activity to discord, if there are any errors resolve the failed promise
  rpc.setActivity(tempArgs).catch(console.error)
}

// finally, execute the setActivity() function only when rpc is ready

rpc.on('ready', () => {
  username = rpc.user.username + '#' + rpc.user.discriminator

  setActivity()

  setInterval(() => { // set an interval of 10 seconds
    check() // check if there's anything new, if so store the object
    setActivity() // set the activity
  }, 10e3) // wait 10 seconds before repeating the same thing
})

function clientUpdate () {
  rpc.destroy() // *properly* close the rpc
  rpc = new DiscordRPC.Client({ transport: 'ipc' }) // create a new one
  rpc.login({ clientId }).catch(console.error) // register with the new client id
  loadImages()
}

ipcRenderer.on('destroy', () => {
  rpc.destroy()
})

// Login to finalize everything
rpc.login({ clientId }).catch((err) => {
	console.log(`${err}`)
	alert('Could not connect to discord')
	change()
}) // again resolve failed promise


src/.old_main.js
// MIT License

// Copyright (c) 2020 theqoobee, qoobes

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
'use strict'

// Including libraries
const { app, BrowserWindow, ipcMain } = require('electron')
const path = require('path')
const url = require('url')
const DiscordRPC = require('discord-rpc')
var write = false
// Set the mode to development
process.env.NODE_ENV = 'development'

// This is to prevent the app from launching lots of times during
// installation on windows
if (require('electron-squirrel-startup')) {
  app.quit()
}

// Define the window

let mainWindow

const createWindow = () => {
  mainWindow = new BrowserWindow({ // Setting the dimnesions
    width: 800,
    height: 450,
    vibrancy: 'dark', // mac only
    webPreferences: {
      nodeIntegration: true // Make sure node is there
    }
  })

  mainWindow.openDevTools()

  mainWindow.removeMenu() // I don't need the menu there

  // Load the correct file
  mainWindow.loadFile(path.join(__dirname, 'index.html'))

  // Make sure the window is gone whne closed
  mainWindow.on('closed', () => {
    mainWindow = null
  })
}

// The function that creates the window

// When the app is loaded, create the new window
app.on('ready', createWindow)

// Instructions on what to do when the window is closed

app.on('window-all-closed', () => {
  // This is because mac allows all windows to be closed with
  // the app still staying open
  if (process.platform !== 'darwin') {
    // app.quit(); // quit the app
    createWindow()
  }
})

app.on('activate', () => {
  // An extension of the code above
  if (mainWindow === null) {
    createWindow()
  }
})
// ------------------------------------------------------------------------------------------------------
// ------------------------------------------------------------------------------------------------------
// The above code was setup, now that that's done let's get down to the juicy parts

// To elaborate how this works, the main.js function acts like an intermediary between renderer.js and rpc.js
// So rpc.js checks for updates every 12 seconds, by seeing if the args object in the main functions changed
// if it has rpc.js updates the rich presence, and if not, nothing changes
// when the user decides to change something in the rpcengine app, renderer beams the new info over to main.js

// THE DISCORD RPC SECTION OF THE CODE

var clientId = '665244956805300274'
var text1, text2, timer, imagelg, imagesm
DiscordRPC.register(clientId)

const rpc = new DiscordRPC.Client({ transport: 'ipc' })

const startTimestamp = new Date()

async function setActivity () {
  if (!rpc || !mainWindow) {

  }
}

ipcMain.on('args', (event, args) => {
  text1 = args.text1
  text2 = args.text2
  timer = args.timer
  imagelg = args.imagelg
  imagesm = args.imagesm
  write = true
  event.returnValue = { success: true, user: username }
})

if (write) {
  console.log(`yolo ${text1}`)
  try {
    if (args.timer) {
      rpc.setActivity({
        details: args.text1,
        state: args.text2,
        startTimestamp,
        largeImageKey: args.imagelg,
        smallImageKey: args.imagesm,
        instance: false
      })
    } else {
      rpc.setActivity({
        details: args.text1,
        state: args.text2,
        largeImageKey: args.imagelg,
        smallImageKey: args.imagesm,
        instance: false
      })
    }
  } catch (err) {
    console.log(`problem with  ${text1}`)
  }
}

var username
rpc.on('ready', () => {
  username = rpc.user.username + '#' + rpc.user.discriminator

  setActivity()
  setInterval(() => {
    setActivity()
  }, 15e3)
})

rpc.login({ clientId }).catch(console.error)


src/js/renderer.js
// MIT License

// Copyright (c) 2020 theqoobee, qoobes

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

const $ = require('jquery') // jquery becuase im too lazy to learn rn
const ipcRenderer = require('electron').ipcRenderer // Used to communicate between renderer and main
var active = false
var changed = false
var defArgs = {
  details: 'Using rpcengine',
  state: 'by theqoobee',
  startTimestamp: true, // Timestamp to track the passage of time
  largeImageKey: 'megu1', // name of the pics
  smallImageKey: 'naruto_sleep', // ^^^
  largeImageText: "I'm big image", // the text displayed when hovering over them
  smallImageText: "I'm small image",
  instance: false
}
var args = defArgs
/* TODO LIST:
 * Make it work as it is with regular vars
 * Dynamically add the selections
 * Sync the html code to work with this
 */

setInterval(() => {
  update()
}, 5e3)

async function update () {
  if (active) {
    args = {
      details: $('#text1').val(),
      state: $('#text2').val(),
      startTimestamp: $('#timer').val() == 'Yes Timer',
      largeImageKey: $('#imagelg').val() == 'No Large Image' ? undefined : $('#imagelg').val().toLowerCase(),
      smallImageKey: $('#imagesm').val() == 'No Large Image' ? undefined : $('#imagesm').val().toLowerCase(),
      largeImageText: $('#imagelgText').val(),
      smallImageText: $('#imagesmText').val(),
      instance: false
    }

    if (args.details.length > 2 && args.state.length > 2) {
      ipcRenderer.send('update', args)
      if (changed) {
        change()
        changed = false
      }
    } else {
      args = defArgs
      log('Please enter at least 2 characters for the first two fields', 'error')
      active = false
    }
  }
}

ipcRenderer.on('username', (event, args) => {
  log(`Successfuly connected for <strong> ${args} </strong>`, 'big')
})

$(document).ready(() => { // When the document is fully loaded
  // The function below changes what happens when the enter key is pressed on all elements
  window.addEventListener('keyup', (event) => {
    if (event.keyCode === 13) {
      event.preventDefault()
      $('#submit').click()
    }
  })

  loadImages() // load the image options as soon as the document is ready
})

function toggleActive () {
  return active = !active // toggles active
}

function enrich () {
  toggleActive()

  update()
  if (active) {
    $('#submit').replaceWith('<button id="submit"onclick="enrich()" class="btn btn-dark">Presence: ON</button>')
  } else {
    $('#submit').replaceWith('<button id="submit"onclick="enrich()" class="btn btn-dark">Presence: OFF</button>')
  }
}

// I'll use this later

// function removeHelp () {
//   $('#helperText').toggleClass('hidden')
// }

// The following are functions used to visually represent what's going on for the user
// change() changes everything to red, for errors

function change () {
  $('#output').toggleClass('bg-danger')
  $('#ico').toggleClass('bg-danger')
  $('input').toggleClass('bg-danger')
  $('select').toggleClass('bg-danger')
  $('input').toggleClass('white')
}

// Log uses the "command prompt" in the UI to tell users what's going on
function log (text, type) {
  if (type === 'error' && !changed) { // Paramater error makes everything turn red
    $('#textOutput').replaceWith(`<p class="text-monospace text-white small" id="textOutput">${text};</p>`)
    change()
    changed = true
  } else if (type === 'big') {
    $('#textOutput').replaceWith(`<p class="text-monospace text-white" id="textOutput">${text};</p>`)
  } else {
    $('#textOutput').replaceWith(`<p class="text-monospace text-white small" id="textOutput">${text};</p>`)
  }
}

// Sends a message to open the settings window
function settingWindow () {
  ipcRenderer.send('settings')
}
// Toggles the topbar menu cause it's annoying
function menu () {
  ipcRenderer.send('menu')
}

// The next bit dynamically loads in the images for the choosers

async function loadImages () {
  const assets = await getAssets(clientId) // Self-explanatory i'd say

  let assetElementsLg = '<select id="imagelg" class="form-control"><option>No Large Image</option>' // The first options
  let assetElementsSm = '<select id="imagesm" class="form-control"><option>No Small Image</option>'
  const namesOfAssets = []

  assets.forEach(asset => {
    namesOfAssets.push(asset.name) // cycle through the assets and make an array of strings, because it can be sorted
  })
  namesOfAssets.sort() // sordit
  namesOfAssets.forEach(asset => {
    assetElementsLg += `<option>${asset}</option>` // add the options as strings
    assetElementsSm += `<option>${asset}</option>`
  })
  assetElementsLg += '</select>' // add the ending 'bracket'
  assetElementsSm += '</select>'
  $('#imagelg').replaceWith(assetElementsLg) // and finally put it in
  $('#imagesm').replaceWith(assetElementsSm)
}

async function getAssets (id) {
  url = `https://discordapp.com/api/v6/oauth2/applications/${id}/assets`

  const assets = await fetch(url)
    .then(response => {
      return response.json()
    })
    .then(data => {
      returns = data
      return data
    })
  if (assets.code !== undefined) {
    return 'INVALID_ID_ERROR'
  } else { return assets }
}


src/js/settings.js
// MIT License

// Copyright (c) 2020 theqoobee, qoobes

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

// This file regulates the settings window
const $ = require('jquery') // jquery becuase im too lazy to learn rn
const ipcRenderer = require('electron').ipcRenderer // Used to communicate between renderer and main
var url, customClientId, returns
// Function for opening the wiki in another browser window
function wiki () {
  require('electron').shell.openExternal('https://github.com/theqoobee/rpcengine/wiki')
}

// save the client id and shoot it to main
async function kel () {
  customClientId = $('#ClientId').val()

  if (await prove(customClientId)) {
    ipcRenderer.send('advanced', customClientId)
    $('#multiSelect').replaceWith('<p class="text-success small ml-3" >Sucessfully set new client id! Wait 10-15 seconds for changes to apply.</p>')
    $('#clientButton').replaceWith('<button class="btn button success" id="clientButton" onclick="exit()">Exit</button>')
    $('#defaultButton').remove()
    $('#helpText').toggleClass('success')
  }
}

function def () { // resets the clientid to default
  ipcRenderer.send('advanced', '665244956805300274')
  $('#multiSelect').replaceWith('<p class="text-success small ml-3" >Sucessfully set new client id! Wait 10-15 seconds for changes to apply.</p>')
  $('#clientButton').replaceWith('<button class="btn button success" id="clientButton" onclick="exit()">Exit</button>')
  $('#helpText').toggleClass('success')
  $('#defaultButton').remove()
}

function exit () {
  ipcRenderer.send('exitSettings')
}

async function prove (id) {
  if (id.length !== 18) {
  	alert('Wrong clientId: too short')
    return false
  }
  if (await getAssets(id) === 'INVALID_ID_ERROR') {
  	alert('Wrong clientId: doesn\'t exist')
    return false
  }
  return true
}

async function getAssets (id) {
  url = `https://discordapp.com/api/v6/oauth2/applications/${id}/assets`

  const assets = await fetch(url)
    .then(response => {
      return response.json()
    })
    .then(data => {
      returns = data
      return data
    })
  if (assets.code !== undefined) {
    	return 'INVALID_ID_ERROR'
  } else { return assets }
}


src/js/old_renderer.js
// //MIT License

// Copyright (c) 2020 theqoobee, qoobes

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

const $ = require('jquery')
const ipcRenderer = require('electron').ipcRenderer
var text1, text2, imagelg, imagesm, timer
let changed = false
let status
var mode = 'Code'

function enrich () {
  if ($('#text1').val().length < 2 || $('#text2').val().length < 2) {
    log('Please enter at least 2 characters', 'error')
    return
  }

  text1 = $('#text1').val()
  text2 = $('#text2').val()
  imagelg = $('#imagelg').val() == 'No Large Image' ? undefined : $('#imagelg').val().toLowerCase()
  imagesm = $('#imagesm').val() == 'No Small Image' ? undefined : $('#imagesm').val().toLowerCase()
  timer = $('#timer').val() == 'Yes Timer'

  status = ipcRenderer.sendSync('args', { text1, text2, imagelg, imagesm, timer })
  if (status.success === true) {
    if (changed) {
      changed = false
      change()
    }
    log(`Successfuly connected for <strong>${status.user}</strong>`)
  } else {
    document.write(
      '<h1>A serious error has occured</h1>',
      '<br> <h3>please report this on the following link </h3>',
      '<br> <a style="font-size: 40px" href="https://github.com/theqoobee/rpcengine/issues">Github.</a>',
      '<br> <h3>Please forward this error too <br> </h3>',
    `<br> <h4 style = 'color: red'> ${status}</h4>`
    )
  }
}

$(document).ready(() => {
  window.addEventListener('keyup', (event) => {
    if (event.keyCode === 13) {
      event.preventDefault()
      $('#submit').click()
    }
  })
})

function change () {
  $('#output').toggleClass('bg-danger')
  $('#ico').toggleClass('bg-danger')
  $('input').toggleClass('bg-danger')
  $('select').toggleClass('bg-danger')
  $('input').toggleClass('white')
}

function log (text, type) {
  $('#textOutput').replaceWith(`<p class="text-monospace text-white" id="textOutput">${text};</p>`)
  if (type === 'error' && !changed) { change(); changed = true };
}

ipcRenderer.on('name', (event, args) => {
  console.log(args)
})


src/js/.old_renderer.js
// MIT License

// Copyright (c) 2020 theqoobee, qoobes

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

const $ = require('jquery')
const ipcRenderer = require('electron').ipcRenderer
var text1, text2, imagelg, imagesm, timer
let changed = false
let status
var mode = 'Code'

function enrich () {
  if ($('#text1').val().length < 2 || $('#text2').val().length < 2) {
    log('Please enter at least 2 characters', 'error')
    return
  }

  text1 = $('#text1').val()
  text2 = $('#text2').val()
  imagelg = $('#imagelg').val() == 'No Large Image' ? undefined : $('#imagelg').val().toLowerCase()
  imagesm = $('#imagesm').val() == 'No Small Image' ? undefined : $('#imagesm').val().toLowerCase()
  timer = $('#timer').val() == 'Yes Timer'

  status = ipcRenderer.sendSync('args', { text1, text2, imagelg, imagesm, timer })
  if (status.success === true) {
    if (changed) {
      changed = false
      change()
    }
    log(`Successfuly connected for <strong>${status.user}</strong>`)
  } else {
    document.write(
      '<h1>A serious error has occured</h1>',
      '<br> <h3>please report this on the following link </h3>',
      '<br> <a style="font-size: 40px" href="https://github.com/theqoobee/rpcengine/issues">Github.</a>',
      '<br> <h3>Please forward this error too <br> </h3>',
    `<br> <h4 style = 'color: red'> ${status}</h4>`
    )
  }
}

$(document).ready(() => {
  window.addEventListener('keyup', (event) => {
    if (event.keyCode === 13) {
      event.preventDefault()
      $('#submit').click()
    }
  })
})

function change () {
  $('#output').toggleClass('bg-danger')
  $('#ico').toggleClass('bg-danger')
  $('input').toggleClass('bg-danger')
  $('select').toggleClass('bg-danger')
  $('input').toggleClass('white')
}

function log (text, type) {
  $('#textOutput').replaceWith(`<p class="text-monospace text-white" id="textOutput">${text};</p>`)
  if (type === 'error' && !changed) { change(); changed = true };
}

ipcRenderer.on('name', (event, args) => {
  console.log(args)
})


src/main.js
// MIT License

// Copyright (c) 2020 theqoobee, qoobes

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
'use strict'

// Including libraries
const { app, BrowserWindow, ipcMain } = require('electron')
const path = require('path')
const url = require('url')
// const rpc = require('./js/rpc.js');
const isEqual = require('lodash.isequal')
var rpcjs
// Set the mode to development
process.env.NODE_ENV = 'production'


var menu = false

// Define the window
let mainWindow

const createWindow = () => {
  mainWindow = new BrowserWindow({ // Setting the dimnesions
    icon: 'https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fcdn2.iconfinder.com%2Fdata%2Ficons%2Fios-7-icons%2F50%2Fengine-512.png&f=1&nofb=1',
    width: 800,
    height: 450,
    vibrancy: 'dark', // mac only
    resizable: false,
    webPreferences: {
      nodeIntegration: true // Make sure node is there
    }
  })

  // mainWindow.openDevTools()

  // // this is temporary

  // if (!menu) {
  mainWindow.removeMenu()
  // }
  // Load the correct file
  mainWindow.loadFile(path.join(__dirname, 'index.html'))

  // Make sure the window is gone whne closed
  mainWindow.on('closed', () => {
    mainWindow = null
  })
}

// The function that creates the window

// When the app is loaded, create the new window
app.on('ready', createWindow)

// Instructions on what to do when the window is closed

app.on('window-all-closed', () => {
  // This is because mac allows all windows to be closed with
  // the app still staying open
  if (process.platform !== 'darwin') {
    rpcjs.send('destroy')
    app.quit() // quit the app
  }
})

app.on('activate', () => {
  // An extension of the code above
  if (mainWindow === null) {
    createWindow()
  }
})

// This is for integrating the special settings

let settingWindow

ipcMain.on('settings', (event) => {
  if (settingWindow === undefined || settingWindow === null) {
    createSettingWindow()
  }
})
function createSettingWindow () {
  settingWindow = new BrowserWindow({ // Setting the dimnesions
    parent: mainWindow,
    width: 600,
    height: 300,
    vibrancy: 'dark', // mac only
    resizable: false,
    webPreferences: {
      nodeIntegration: true // Make sure node is there
    }
  })

  // settingWindow.openDevTools()

  settingWindow.removeMenu() // I don't need the menu there, it's ew.
  // if (!menu) {
  // settingWindow.removeMenu()
  // }// this is again temporary

  // Load the correct file
  settingWindow.loadFile(path.join(__dirname, 'settings.html'))

  // Make sure the window is gone whne closed
  settingWindow.on('closed', () => {
    settingWindow = null
  })
}
// ------------------------------------------------------------------------------------------------------
// ------------------------------------------------------------------------------------------------------
// The above code was setup, now that that's done let's get down to the juicy parts

// To elaborate how this works, the main.js function acts like an intermediary between renderer.js and rpc.js
// So rpc.js checks for updates every 12 seconds, by seeing if the args object in the main functions changed
// if it has rpc.js updates the rich presence, and if not, nothing changes
// // when the user decides to change something in the rpcengine app, renderer beams the new info over to main.js

var changed = { args: false, clientId: false } // Two variables to keep track of changed
var customClientId
var username
var mainArgs
var argsReturnValue, clientReturnValue

// This function reacts to a message from the renderer process, specifically looking at if anything changed
// If anything changed, it give it to the rpc function wheen it checks
ipcMain.on('update', (event, args) => {
  if (isEqual(mainArgs, args) === false) {
    changed.args = true
    mainArgs = args
  }
  event.sender.send('username', username)
})

// Reacts to rpc checking if there's anything new
ipcMain.on('check', (event, args) => {
  username = args // we passed the username from the rpc.js function
  // check for new general args
  if (changed.args !== false) {
    argsReturnValue = mainArgs
    changed.args = false
  } else { argsReturnValue = false }

  // check for new clientid
  if (changed.clientId === true) {
    clientReturnValue = customClientId
    changed.clientId = false
  } else { clientReturnValue = false }

  event.returnValue = { args: argsReturnValue, clientId: clientReturnValue }
})

ipcMain.on('advanced', (event, args) => {
  changed.clientId = true
  customClientId = args
})

ipcMain.on('info', (event) => {
  rpcjs = event.sender
})

ipcMain.on('exitSettings', (event) => {
  if (settingWindow !== null) {
    settingWindow.close()
  }
})

// // To be removed for production
// // This just toggles the menu so i don't have to restart the app every time i want to see what it looks like without it
// ipcMain.on('menu', (event) => {
//   menu = !menu
//   mainWindow.close()
//   if (settingWindow) {
//     settingWindow.close()
//     createSettingWindow()
//   }
//   // createWindow()
// }) // THIS PART REALLY NEEDS REMOVAL LATER ^^^^^
